File: Namrata-Sanger/131. Palindrome Partitioning.js
/*
Day 30
131. Palindrome Partitioning
Link: https://leetcode.com/problems/palindrome-partitioning/submissions/
Level: Medium
*/

var partition = function(s) {
    const ret = [], path = [];
    
    const isPalindrome = (i, j, s) =>{
        while(i < j - 1){
            if(s[i] !== s[j-1]){
                return false;
            }
            ++i;
            --j;
        }
        return true;
    }
    
    const backtrack = (i, j, s) =>{
        if(i === s.length){
            ret.push([...path]);
            return;
        }
        
        for(let index = j; index <= s.length; ++index){
            if(isPalindrome(i, index, s)){
                path.push(s.slice(i, index));
                backtrack(index, index+1, s);
                path.pop();
            }
        }
    };
    
    backtrack(0, 1, s);
    return ret;
};

File: Vindeep-Chaudhari/131_Palindrome_Partitioning.cpp
/**
 * Problem Link - https://leetcode.com/problems/palindrome-partitioning/
 */

class Solution {
public:
    vector<vector<string>> ans;
    vector<vector<string>> partition(string s) {
        if(s.empty()) return ans;
        
        vector<string> path;
        dfs(0, s, path);
        
        return ans;
    }
    
    void dfs(int index, string& s, vector<string>& path) {
        if(index == s.size()) {
            ans.push_back(path);
            return;
        }
        for(int i = index; i < s.size(); ++i) {
            if(isPalindrome(s, index, i)) {
                path.push_back(s.substr(index, i - index + 1));
                dfs(i+1, s, path);
                path.pop_back();
            }
        }
    }
    
    bool isPalindrome(string& s, int start, int end) {
        while(start <= end) {
            if(s[start++] != s[end--])
                return false;
        }
        return true;
    }
};